content/blog/cran-checks-compiled-code/index.Rmd

```diff
@@ -0,0 +1,225 @@
+-   For users with R <= 3.5 on a system with an older compiler, package authors can instruct users to edit their `~/.R/Makevars` file to include this line: `CXX_STD=CXX11`.
+
+The tidyverse has a [policy of supporting four previous versions](https://www.tidyverse.org/blog/2019/04/r-version-support/) of R. Currently that includes R 3.5, but with the upcoming release of R 4.3 the minimum version we will support is R 3.6. As we won't be supporting R 3.5 in the near future, you should not feel pressured to either unless you have a compelling reason.
+
+## WARNING regarding the use of <code>sprintf()</code> in C/C++
+
+Another recent change in CRAN checks on R-devel that authors might encounter is the disallowing of the use of the C functions <code>sprintf()</code> and `vsprintf()`. R CMD check on R-devel may throw warnings that look something like this:
+
+```
+checking compiled code ... WARNING
+File 'fs/libs/fs.so':
+  Found 'sprintf', possibly from 'sprintf' (C)
+    Object: 'file.o'
+Compiled code should not call entry points which might 
+terminate R nor write to stdout/stderr instead of to the 
+console, nor use Fortran I/O nor system RNGs nor [v]sprintf.
+See 'Writing portable packages' in the 'Writing R Extensions' manual.
+```
+
+According to the [NEWS for R-devel](https://developer.r-project.org/blosxom.cgi/R-devel/NEWS/2022/12/24#n2022-12-24) (which will be R 4.3):
+
+> The use of sprintf and vsprintf from C/C++ has been deprecated in macOS 13 and is a known security risk. R CMD check now reports (on all platforms) if their use is found in compiled code: replace by snprintf or vsnprintf respectively.
+
+These are considered to be a security risk because they potentially allow [buffer overflows](https://en.wikipedia.org/wiki/Buffer_overflow) that write more bytes than are available in the output buffer. This is a risk if the text that is being passed to <code>sprintf()</code> comes from an uncontrolled source.
+
+Here is a very simple example:
+
+```{r}
+library(cpp11)
+
+cpp_function('
+  int say_height(int height) {
+    // "My height is xxx cm" is 19 characters but we need
+    // to add one for the null-terminator
+    char out[19 + 1];
+    int n;
+    n = sprintf(out, "My height is %i cm", height);
+    Rprintf(out);
+    return n;
+  }
+'
+)
+
+say_height(182)
+```
+
+```{r}
+#| eval: false
+say_height(1824) # This will abort due to buffer overflow
+```
+
+### How to fix it
+
+In most cases, this should be a simple fix: replace <code>sprintf()</code> with `snprintf()` and `vsprintf()` with `vsnprintf()`. These `n` variants take a second parameter that specifies the maximum number of bytes written. If the output is a static buffer, you can use `sizeof()`:
```

Comment:
Maybe also explain the return value of `snprintf()`? Since it does come up below.